test/parallel/test-fs-write-stream-partial-write.js

```diff
@@ -0,0 +1,40 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const path = require('path');
+const fs = require('fs');
+
+const tmpdir = require('../common/tmpdir');
+const file = path.join(tmpdir.path, 'out.txt');
+
+tmpdir.refresh();
+
+{
+  const _fs_write = fs.write;
+  let i = 0;
+  fs.write = function(fd, data, offset, length, pos, cb) {
```

Comment:
I'm not a fan of relying on the fact that `createWriteStream` uses the public `fs.write` function, unless we already do so elsewhere in the test suite. Is there a legitimate case in an operating system where writes are reliably partial that we can use to test this? Maybe some system file on Linux? or TCP sockets?

Reply:
I wrote the test based on some existing tests (https://github.com/nodejs/node/blob/master/test/parallel/test-fs-write-stream-change-open.js#L35) that rely on intercepting public `fs` APIs. Monkey patching `fs.write()` is the easiest way to test this IMO, since partial writes usually occur non-deterministically.
Setting up the test with actual infrastructure is tricky because it is entirely up to the operating system to decide if it can wholly fulfill the write operation.